### PR TITLE
Initial gas metering

### LINKED ISSUE #1: Add Gas Metering
Add a gasLimit parameter when a program is executed.
- The EVM should return an outOfGas response if the program consumed to much gas.
- On each step the remaining gas should be decreased by the cost of the specific OP-code.

Check https://github.com/ethereum/py-evm for reference.
Check the [Architecture docs](https://github.com/parsec-labs/solevm-truffle/blob/master/docs/Architecture.md) for more details.

### PR DESCRIPTION
**Pull request by pinkiebell**
Fixes #1 Depends on #5

### CODE REVIEW

File: contracts/EthereumRuntime.sol [hunk 1]
```diff
@@ -65,6 +65,7 @@ contract IEthereumRuntime is EVMConstants {
         bytes accountsCode;
         uint[] logs;
         bytes logsData;
+        uint64 gasRemaining;
```
Comment (pinkiebell) [inline contracts/EthereumRuntime.sol]: Should this stay or be removed, changed to account for gasUsed?

Comment (pinkiebell) [reply contracts/EthereumRuntime.sol]: @johannbarbie ?

**Comment by troggy** [comment]
some of the tests are failing for me, like `BLOCKHASH` opcode test. It doesn't fail on my branch. Can you check please? Maybe some merge error? Ignore `GAS` opcode test, it is not correct — working on it

**Comment by troggy** [comment]
as per tests gas is not reduced for the following opcodes: PC, JUMP, JUMPI, CODESIZE, BLOCKHASH.

The first three have special treatment in the code, so maybe there is an error indeed. For the last two it maybe a problem with tests, cause they are simple once — will check the tests

**Comment by troggy** [comment]
One more thing: your PR is now sharing a bunch of commits with #5. I'm a bit worried this may cause merge problems. Please consider rebasing this PR on top of `feature/off-chain-execution` branch if possible

**Comment by pinkiebell** [comment]
Thanks for the review! I will resolve all issues :+1

**Comment by pinkiebell** [comment]
@troggy 
Let me know that you think of my last commit.

**Pull request merged**